File: training/csharp/src/De.Sots.CellarsAndCentaurs/Domain/Model/Dice.cs
using System;

namespace De.Sots.CellarsAndCentaurs.Domain.Model;

public static class Dice
{
    private static readonly Random Random = new();

    public static int RollD20() => Roll(20).Value;

    public static DiceRoll Roll(int sides) => new(sides, Random.Next(1, sides + 1));
}

public readonly record struct DiceRoll(int Sides, int Value)
{
    public bool IsCritical => Value == Sides;
}

File: sample-projects/ruby/lib/de/sots/cellars_and_centaurs/domain/model/dice.rb
# frozen_string_literal: true

require_relative '../../version'

module De::Sots::CellarsAndCentaurs::Domain::Model
  class Dice
    attr_reader :sides

    def initialize(sides)
      @sides = sides
    end

    def roll
      DiceRoll.new(self, rand(1..sides))
    end
  end

  class DiceRoll
    attr_reader :dice, :value

    def initialize(dice, value)
      @dice = dice
      @value = value
    end

    def critical?
      value == dice.sides
    end
  end

  D20 = Dice.new(20)

  def self.roll_d20
    d20Roll = D20.roll
    d20Roll
  end
end
